### ISSUE TITLE
App stops working in background on Android 11

### ISSUE THREAD

**Issue by self-related** [TylerWilliamson/QuickWeather#73]
App stops working in background on my Android 11 devices every time after a couple of hours (used on LineageOS and Android One devices), same results even with "not optimize battery" set. I also use Gadgetbridge and it works well in background and never gets killed, so I guess it might be a bug.
Both devices have 3 Gb RAM.

**Comment by TylerWilliamson** [comment]
Thank you for reporting this issue. I have not experienced this issue on my devices, but I have seen other issues reported on LineageOS devices.

A few initial questions:

1. Are you using a set location or "My Location"?
2. Are you using the "Current Weather" notification? It should have a "last updated" date on it. Does it too stop updating?
3. Do you restrict internet access in the background? QuickWeather needs internet access in the background
4. Do you let your phone "doze"? If so, the phone will stay asleep and not wake up to update Gadgetbridge until you unlock the phone
5. Are there any other battery / privacy / optimization settings that you have active that you feel may contribute?

**Comment by self-related** [comment]
1. I used "My location" (gonna try manual location instead)
2. Yes, it updates 3-4 times (as I noticed in notification history), then the notification just disappear
3. No restrictions
4. Yes, but it stops working in background during normal use 
5. I think I don't use any optimization. But I also get "error loading radar images" every time I open the app. upd: Now I tried with VPN and had no loading error, but it looks different now and doesn't show weather alerts. 

![error](https://github.com/TylerWilliamson/QuickWeather/assets/105994362/862e7721-6dd0-42aa-a222-175162165068)
![no-error](https://github.com/TylerWilliamson/QuickWeather/assets/105994362/fb4952ec-c04f-4257-8ea5-2030b49fc67f)

**Comment by TylerWilliamson** [comment]
Could you let me know the manufacturer(s) of your device(s) that experienced this issue? If it is xiaomi, oppo, vivo, letv, or honor, there seems to be an "autostart" permission in the security app that needs to be granted to QuickWeather.

**Comment by self-related** [comment]
4.The notification just accidentally stops being displayed

I used with Redmi Note 7 on LOS 18 and autostart was granted (and it started successfully after reboot). 
The other device is https://www.devicespecifications.com/en/model/d7fc59eb and it has "App standby optimizer" in battery settings, I've set it to "Not Optimize" and it didn't solve the problem.

I also set "Allow all the time" for location permission, but I've never noticed the "location icon" at the top of the screen while the notification was still working in background (as the icon appears when I update my location in other apps). My guess it might crash after failing to update location on my devices. Thank you for the attention, I'll try a manual location and will report the result.

**Comment by self-related** [comment]
Okay, the notification has disappeared again, manual location was used. Maybe it's my OS oom killer or smth, but again it never kills gadgetbridge. I'd be glad to provide more info for investigation.

**Comment by TylerWilliamson** [comment]
Thanks for the info. I'll dig into how Gadgetbridge does its background work. It most likely is using a constantly running service, whereas QuickWeather utilizes "WorkManager", which creates a small job every 15 minutes to update the weather. In theory, this reduces the battery utilization of QuickWeather in the background. Apparently though, the implementation of WorkManager is device and OS specific, which is where the issues can occur

**Comment by TylerWilliamson** [comment]
I did a bit of digging and Gadgetbridge does seem to use a constantly running service. This probably keeps the OS from killing it. As for your devices, since they are running modern OS's with fairly low RAM, it does not leave much room for apps to run and Xiaomi seems to be pretty aggressive with killing apps in the background. I'm unsure if LineageOS affects it. See [this site](https://dontkillmyapp.com/xiaomi) for more info.

If you are able to capture a bug report and are comfortable with sending it, there may be clues in there if this issue is with the OS/device or with QuickWeather.

I'll likely be adding an "Advanced" settings menu where I can add log files, etc to assist with the debug process in later updates

**Comment by self-related** [comment]
I captured a bug report from my current device. I think I played with those optimization settings a little, so 2023-07-13 20:00 should be the time when everything done correctly.
[bugreport-BQru-6868L-RP1A.201005.001-2023-07-13-23-58-29.zip](https://github.com/TylerWilliamson/QuickWeather/files/12044672/bugreport-BQru-6868L-RP1A.201005.001-2023-07-13-23-58-29.zip)

**Comment by TylerWilliamson** [comment]
Thank you for the bug report. Unfortunately, I wasn't able to find any clues to why the job is getting killed in the background. I will leave this issue open as I add logging features and I may ask for updates after those features are in place. If you find any information, please feel free to post it here.

**Comment by TylerWilliamson** [comment]
I am seeing this issue on my own device, and others are seeing it as well. I am working on a solution and will update this thread when I make the fix

**Comment by firecat53** [comment]
Version 2.3.3 from f-droid works correctly. It looks like something changed with 2.3.4, although nothing in the change log suggested anything obvious.

**Comment by TylerWilliamson** [comment]
This error started occurring before 2.3.4. It seems to be an intermittent issue and I think I may know a solution. I'll test it tonight and if it seems to work, I can put an APK here for testing

**Comment by TylerWilliamson** [comment]
I believe I have found the issue. I was cancelling and re-enqueuing the background job to reset the 15 minute timer whenever the app was opened, but the background job also was calling that logic. In the past, this did not cause an issue, but I think there was a change in the behavior of WorkManager in its newer releases. Regardless, I have removed the "cancel" part of the logic, so the timer will not reset and the background job will run every 15mins. This should resolve the issue. I am testing now, and if anyone in this thread would also like to test, I have uploaded a zipped preview APK. Please note that the APK has a different Application ID so you can install it alongside the version you have installed from F-Droid, Google Play Store, etc.

[QuickWeather-2.3.5-PREVIEW.zip](https://github.com/TylerWilliamson/QuickWeather/files/12136845/QuickWeather-2.3.5-PREVIEW.zip)

**Comment by TylerWilliamson** [comment]
Unfortunately, the change above did not resolve the issue. It turns out the background worker is running every 15 minutes, successfully. The issue is a bit more complicated.

There is [an Android bug](https://issuetracker.google.com/issues/37124582) which affects the Day/Night mode. Basically, to have the correct day/night mode and colors, you need to create a WebView. This code is in the TylerUtils library, in my ColorUtils utility. I call that method when I instantiate my ColorHelper. This class was previously instantiated when the application is instantiated as part of the QuickWeather application class's OnCreate. I changed the ColorHelper class to be a more efficient singleton instead of a utility class, so it is now instantiated whenever it is first needed, in this case, in WeatherWorker. When the singleton is instantiated, it throws an error "WebView cannot be initialized on a thread that has no Looper" because it's creating a View in the background in the Worker. If we do this instead in the application class, we should be fine. I am testing now, and there is a new preview APK if anyone wants to test with me.

TL;DR: Previous change didn't fix the issue, I have a new solution. Preview APK attached.

[QuickWeather-2.3.5-PREVIEW.apk.zip](https://github.com/TylerWilliamson/QuickWeather/files/12139979/QuickWeather-2.3.5-PREVIEW.apk.zip)

**Comment by TylerWilliamson** [comment]
The fix has been released in version [2.3.5](https://github.com/TylerWilliamson/QuickWeather/releases/tag/2.3.5). If the error persists, please let me know in this thread and I will reopen the issue.
